### PR TITLE
Servicelocator tree traversal

### PR DESCRIPTION
**Pull request by SamMousa**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes

The traversal already works, but checking the tree didn't. I've added more tests and implemented traversal for searching.

### CODE REVIEW

**Review by dmirogin — APPROVED** [review]

File: framework/base/Module.php [hunk 1]
```diff
@@ -729,4 +729,21 @@ public function get($id, $throwException = true)
...
+
+    /**
+     * @inheritdoc
+     * @param string $id
+     * @param bool $checkInstance
+     * @return bool
+     * Since version 2.0.13, if a component isn't defined in the module, it will be looked up in the parent module.
+     * The parent module may be the application.
+     */
+    public function has($id, $checkInstance = false)
```
Comment (samdark) [inline framework/base/Module.php]: How about the following variant? Looks a bit more readable to me...

```php
if (parent::has($id, $checkInstance)) {
    return true;
}

if (isset($this->module)) {
    return $this->module->has($id, $checkInstance);
}

return false;
```

Comment (SamMousa) [reply framework/base/Module.php]: I've changed it to a one liner that feels more readable; i'm not a fan of using `if... return true;`.

Comment (samdark) [reply framework/base/Module.php]: Well, my variant is close in reading to description. Yours is not and it duplicates a single condition multiple times.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
@yiisoft/reviewers please take a look.

File: framework/base/Module.php [hunk 2]
```diff
@@ -729,4 +729,17 @@ public function get($id, $throwException = true)
...
         return $component;
     }
+
+    /**
+     * @inheritdoc
+     * @param string $id
+     * @param bool $checkInstance
+     * @return bool
+     * Since version 2.0.13, if a component isn't defined in the module, it will be looked up in the parent module.
+     * The parent module may be the application.
```
Comment (rob006) [inline framework/base/Module.php]: You should remove `@return` and `@param` annotations - they will be inherited from parent class.

File: framework/base/Module.php (hunk 1 above)
Comment (samdark) [reply framework/base/Module.php]: OK. New one looks fine.

File: framework/base/Module.php (hunk 2 above)
Comment (cebe) [reply framework/base/Module.php]: done.

**Review by cebe — APPROVED** [review]

**Comment by SilverFire** [comment]
I'd even mention this change in UPGRADE.md, because it changes previous behavior.

**Review by SilverFire — APPROVED** [review]
I'd even mention this change in UPGRADE.md, because it changes previous behavior.

**Comment by SilverFire** [comment]
Done

**Comment by SilverFire** [comment]
Thinking about method `hasOwn($id)` to leave previous behavior available at least for `has()` method, but can not find a good example to prove its necessity. Do you have any thoughts?

**Comment by SamMousa** [comment]
@SilverFire I think if you need `hasOwn()` as a public method you (very likely) have a violation of separation of concerns.
The module is a `ServiceLocator`, it is not relevant how it gets the services that it locates. For all a consumer knows it could be magically creating objects from thin air. `hasOwn()` therefore should not be necessary.

**Comment by samdark** [comment]
I agree with @SamMousa on this one.

**Comment by samdark** [comment]
Merged. Thank you!

**Comment by pavlm** [comment]
This change broke some code. Particularly when using Podium forum module with customized module components. Podium uses getter methods to access components but new `->has()` method makes them invisible.

**Comment by SamMousa** [comment]
@pavlm Please create a new issue and tag me in it and I'll have a look. Please be more specific and provide a detailed example that I can look at.

**Pull request merged**